Create proofread.md (#1)
slouc/concurrent-effects: proofread.md

- [References](#references)

## Introduction

First, some useful definitions and descriptions:
  
- Blocking:  
  Thread that executes a blocking task will wait on action untill success or failure.
  
- Non-blocking:  
  Thread that executes a non-blocking task will initiate it and immediately continue with another task without waiting. When the first task is done, its result may or may not be processed by the same thread that initiated it.
  
- Synchronous:  
  Thread will complete the task, either by success or failure, before reaching any line after it. This often involves blocking, because task N+1 cannot continue until task N has finished, which might include waiting for the result from another thread.
  
- Asynchronous:  
  Task was started by one thread, but another thread (either logical or physical) will complete the task, and then return the result using a callback. Task N+1 can continue even if N hasn't finished yet.

- Concurrency:  
  State in which there are multiple logical threads of control, whose effects are interleaved.
  
- Parallelism:  
  State in which computations are physically performed in parallel on separate CPU cores.

> 

Reviewer: Great suggestions, I will address them in my next PR.

> Note that concurrency isn't the same as parallelism: we could have concurrency without parallelism (effects are interleaved, but everything is done by one single CPU core), and we could have parallelism without concurrency (multiple cores are running multiple standalone threads that don't interleave). In this text, we are interested in the concurrent aspect of our programs, and we don't care whether some of the tasks are done in parallel or not.

Reviewer: Yes, that's my understanding. Two threads running on two separate CPU cores, one calculating Pi and the other mining bitcoin, are an example of parallelism, but not concurrency.

Also, two threads running on a single core or two fibers running on a single thread only implies concurrency if they are running interleaved. If they run one after another, there is no concurrency.

> Note that concurrency isn't the same as parallelism: we could have concurrency without parallelism (effects are interleaved, but everything is done by one single CPU core), and we could have parallelism without concurrency (multiple cores are running multiple standalone threads that don't interleave). In this text, we are interested in the concurrent aspect of our programs, and we don't care whether some of the tasks are done in parallel or not.

Author: Nice! So it implies:
- programming model
    - synchronous (execution in "linear" flow, line by line) 
    - asynchronous (using constructs, that allow you to run tasks concurrently/ in parallel)
- execution 
    - sequential (no interleaving, tasks are executed one after another) 
    - concurrent (tasks are interleaved)
- implementation of execution 
    - parallel (multiple tasks executed at the same time) 
    - not parallel (either interleaved - implies concurrent execution, or not interleaved - implies sequential execution)

> Note that concurrency isn't the same as parallelism: we could have concurrency without parallelism (effects are interleaved, but everything is done by one single CPU core), and we could have parallelism without concurrency (multiple cores are running multiple standalone threads that don't interleave). In this text, we are interested in the concurrent aspect of our programs, and we don't care whether some of the tasks are done in parallel or not.

Reviewer: I would agree with this, yes (my understanding could be flawed too, though 😅 ).

With a minor correction - asynchronous model means "I can continue executing the rest of the program while this thing here finishes", and then you react when the result is available (you can either express this how we do, via map / flatMap chains which apply the function when the result becomes available, or via callbacks). But I don't think asynchronous model should in any way be connected with parallelism. Yes, async model allows you to run tasks concurrently, but it's your multicore CPU and parallelizable program (e.g. having two or more unrelated tasks) that gives you parallelism. Not async model.

> Also, there seems to be a lot of confusion around blocking vs synchronous and non-blocking vs asynchronous. Some sources (especially ones related to NodeJS) often assume synchronous = blocking and asynchronous = non-blocking, while other sources point out that blocking is always synchronous, but synchronous doesn't always mean blocking (spin lock mechanisms etc.). For all intents and purposes, this text assumes that all synchronous tasks are blocking, and all asynchronous tasks are non-blocking. 

Author: Addressed in #2

## Asynchronous boundary

Concurrency means that a single logical thread can have its tasks distributed across different threads. Or, from the thread perspective, that a single thread can execute interleaved tasks coming from different logical threads.

In order to be able to do so, we have to use asynchronous operations. Once the task has been initiated, it crosses the *asynchronous boundary*, and it resumes somewhere else.

> There are a couple of somewhat famous quotes circulating around:

Author: ```suggestion
There are a couple of somewhat famous quotes circulating:
```

that's probably the same ;)

Reviewer: Or how about simply "There are a couple of somewhat famous quotes" 🙂

Author: the simplified version ("There are a couple of somewhat famous quotes") works best 👍 
addressed https://github.com/slouc/concurrent-effects/pull/2

> "A Future represents a value, detached from time" - Viktor Klang ([link](https://monix.io/docs/2x/eval/task.html#comparison-with-scalas-future))

> "Asynchronous process is a process that continues its execution in a different place or time than the one it started in" - Fabio Labella ([link](https://www.youtube.com/watch?v=x5_MmZVLiSM&t=6m35s))

> "A logical thread offers a synchronous interface to an asynchronous process" - Fabio Labella ([link](https://www.youtube.com/watch?v=x5_MmZVLiSM&t=11m12s))

Bear in mind that saying "synchronous interface" comes from the fact that all our code is basically synchronous. We write commands one after another and they are executed one after another. So when you see a term "synchronous code", pay attention to the context - maybe someone meant "code that doesn't use asynchronous effects (e.g. it blocks on every asynchronous call)". In this text, however, "synchronous code" means just any code, because all code we write is in nature synchronous. It is via usage of asynchronous effects in our code that we get to cross the asynchronous boundary and model asynchronously executed operations.

> All of the above quotes revolve around the same concept of asynchronous boundary; after our computation passes that boundary, we cannot be sure *when* is it going to be finished, which is why we should not block while waiting for it, and we're also not sure *where* is it going to be finished - another OS thread, another node in our phyisical network, somewhere in the cloud etc. This text deals with the details of execution on OS threads (hence the relevance of the third quote), and it will not touch upon any other scenario of asynchronous execution, such as distributing the tasks over nodes in a cluster.

Author: just a typo phyisical -> physical

> All of the above quotes revolve around the same concept of asynchronous boundary; after our computation passes that boundary, we cannot be sure *when* is it going to be finished, which is why we should not block while waiting for it, and we're also not sure *where* is it going to be finished - another OS thread, another node in our phyisical network, somewhere in the cloud etc. This text deals with the details of execution on OS threads (hence the relevance of the third quote), and it will not touch upon any other scenario of asynchronous execution, such as distributing the tasks over nodes in a cluster.

Reviewer: Can you suggest the typo fix that I could then easily commit here, and I will re-write the paragraph in my own PR?

> All of the above quotes revolve around the same concept of asynchronous boundary; after our computation passes that boundary, we cannot be sure *when* is it going to be finished, which is why we should not block while waiting for it, and we're also not sure *where* is it going to be finished - another OS thread, another node in our phyisical network, somewhere in the cloud etc. This text deals with the details of execution on OS threads (hence the relevance of the third quote), and it will not touch upon any other scenario of asynchronous execution, such as distributing the tasks over nodes in a cluster.

Author: Sure!
Addressed: https://github.com/slouc/concurrent-effects/pull/2

When we take a look at how each task is executed by OS thread(s), we can see the importance of concurrency. 

![sync-async](sync-async-boundary.png)

After crossing the asynchronous boundary, the tasks get interleaved across threads in the thread pool. Some of them might even get executed on some thread from another thread pool. This is where the property of concurrency comes into play.

## Threading 

### Threads and thread pools

Threads in JVM map 1:1 to the operating system’s native threads. 

> When CPU stops executing one thread and starts executing another thread, the OS needs to store state of earlier task and restore the state for the current one. 

Author: maybe something like this?

needs to store **the** state of **the** earlier
```suggestion
When CPU stops executing one thread and starts executing another thread, the OS needs to store the state of the earlier task and restore the state for the current one. 
```

Reviewer: Definitely!

- Cooperative scheduling (or cooperative yielding):  
  Tasks suspend themselves, meaning that the currently running task at some point voluntarily suspends its own execution, so that the scheduler can give the thread to other tasks

The role of the scheduler is played by the `ExecutionContext` ([docs](https://docs.scala-lang.org/overviews/core/futures.html#execution-context)). Every `ExecutionContext` schedules threads within one assigned thread pool.

In Scala, there is one global `ExecutionContext`, available as `ExecutionContext.global`. 
Global EC is backed by the `ForkJoinPool`.

Whichever underlying Java executor you rely on, your level of granularity is going to be threads. 
Threads don't cooperate. 
They execute their given set of commands, and the operating system makes sure that they all get some chunk of the CPU time. That's why I can be typing this text, listening to music and compiling some code, all at the same time.

So, in order to allow tasks submitted to the `ExecutionContext` to use the principle of cooperative yielding, we have to explore the concept of **fibers**.
Later we will also look at cooperative yielding [into more detail](#io).

> ## Fibers

Author: I absolutely love the article so far, and will have more questions!

Maybe at this point it'd be good to introduce the concept of green threads, and then mention, that Fibers are one of the implementations of this idea?

Reviewer: I mentioned that in one of the last sections ("a look outside the window") but we could / should mention it here as well

Reviewer: Fixed, and also renamed the window section to "Fibers outside of Scala" which now includes Project Loom and green threads sections.

Note that in some contexts / languages fibers are also known as *coroutines*; "fiber" is usually used in the system-level context, while coroutines are used in the language-level context. 
However, in Scala "fiber" is the preferred term.

Unlike OS and JVM threads which are managed in the kernel space, fibers are managed in the user space.

Fibers map to CPU / JVM threads many-to-few, similarly to how threads map to processes. 
Multiple fibers can run on multiple thread pools, or on the same set of threads from one thread pool, or even on the same thread.
In the case of a single thread, they will take turns executing their code using the available thread. 
Depending on your code, as well as the library that you are using, you can decide to have them cooperate more often, thus achieving fairness, or to cooperate less often, thus sacrificing some fairness for more throughput (concepts of fairness and throughput have been [introduced earlier](##threads-and-thread-pools)).

Fibers are much more lightweight: they consume much less memory, have growable and shrinkable stacks, and can be garbage collected. 
Also, blocking a fiber doesn't block the underlying thread. 

> As a consequence of all that, we don't have to be as careful when creating new fibers as we must when spawning new threads.

Author: Nice! 
maybe... :

```suggestion
As a consequence of all that, we don't have to be as careful when creating new fibers as the upper limit is denoted by available memory. On the other hand, when using threads, we're primarily constrained by the number of cores, but also other aspects.
```

Author: btw. I'm curious to see the mem footprint for 1000 threads (should be around 1GB, given 1thread is at least 1MB) vs 1000 fibers

Reviewer: Thanks, great suggestion! Regarding the comparison, it would definitely be interesting. We could have a single fiber on a single thread, then 1000 fibers on the thread, then 1000 threads without fibers. And watch the memory in each scenario.

This also means that they might have some minor differences in implementation across different libraries (e.g. [cats-effect](https://typelevel.org/cats-effect/datatypes/fiber.html) vs [ZIO](https://zio.dev/docs/overview/overview_basic_concurrency)).

In cats-effect, fiber is a construct with `cancel` and `join`:
```
trait Fiber[F[_], A] {
  def cancel: F[Unit]
  def join: F[A]
}
```

Joining a fiber can be thought of as blocking for completion, but only on a semantic level. 
Remember, blocking a fiber doesn't really block the underlying thread, since it can keep running other fibers. 

## Cats IO type

Type `IO` ([docs](https://typelevel.org/cats-effect/datatypes/io.html)) is used for encoding side effects as pure values. In other words, it allows us to model the operations from the other side of asynchronous boundary in our synchronous code. 

> `IO` serves as an FFI - a [Foreign Function Interface](https://en.wikipedia.org/wiki/Foreign_function_interface). Most common usage of the term FFI is to serve as a translation layer between different programming languages, but `IO` translates side-effectful Scala methods into pure, referentially-transparent values.

Author: 🏅 awesome definition

Reviewer: 🤝

Here are three main ways of modeling asynchronous processes in `IO`:

> ```

Author: ```suggestion
```scala
```

Reviewer: I'll add this to all code snippets.

> def delay[A](body: => A): IO[A] // same as apply()

Author: given 
> "Here are three main ways of modeling asynchronous processes in `IO`:"

 isn't `def delay[A](body: => A): IO[A]` used to just suspend some synchronous computation? ;)

Reviewer: True, it's a mistake in the wording. 👍 I'll change it to something like "main ways of modeling effects".

Btw I want to add a section about cancelling, so this will most likely become four ways, with added `def cancelable[A](k: (Either[Throwable, A] => Unit) => CancelToken[IO]): IO[A]`

def suspend[A](thunk: => IO[A]): IO[A] 

> def async[A](k: (Either[Throwable, A] => Unit) => Unit): IO[A] 

Author: Do you maybe know why `async` signature looks like this?
```scala
def async[A](k: (Either[Throwable, A] => Unit) => Unit): IO[A] 
```
I know it's related to continuations, but doesn't really have the full context. If you know, maybe we can describe it somewhere here as well?

Reviewer: There's a whole section about it later, but I'm not sure if I did a good job explaining it.

```

> We will be using `delay / apply`, in the form of `IO(value)` which desugares into `IO.apply(value)`. 

Author: ```suggestion
We will be using `delay / apply`, in the form of `IO(value)` which desugars into `IO.apply(value)`. 
```

Now let's see what happens if we want to run some `IO` on a separate fiber:

```
...
val program2 = for {
  fiber <- io(1).start
  _ <- io(2)
  _ <- fiber.join
} yield ExitCode.Success
...
```

We define a chain of `IO`s, and then at some point we run some part of that chain on a separate fiber. That's the only difference from the previous program - parts where we `start` and `join` the fiber. Invoking `io1.start` produces an `IO[Fiber[IO, Unit]]`, which means we get a handle over the new fiber which we can then join later, or cancel it on error, or keep it running until some external mechanism tells us to cancel it, etc.

It's important to realize which exact instructions in the example above get executed on which fiber. After we started the execution of `io1` on a separate fiber, everything we did afterwards was done in parallel to the original fiber. We say that the code captured in `io1` was the *source* for the new fiber.

As a small exercise, try adding a small sleep to method `io`:

```
def io(i: Int): IO[Unit] = IO({
  Thread.sleep(3000)
  println(s"Hi from $i!")
})
```

> If you now measure the execution time between `program1` and `program2`, you will see that `program1` runs in six seconds, while `program2` runs in slightly over three seconds. You can find this code in the repository.

Author: ```suggestion
If you now measure the execution time between `program1` and `program2`, you will see that `program1` runs in six seconds, while `program2` runs in slightly over three seconds. You can find this code [in the repository](https://github.com/slouc/concurrent-effects/blob/master/src/main/scala/RunWithFiber.scala).
```

Reviewer: If you want to do this for every time I mention code repository, I would be mega happy! 😂

## Continuations

> In the previous section, we have seen that fiber runs a set of instructions (the source). Any `IO` can, as long as there is a `ContextShift` type class instance available for it, be run in a fiber. This is done by calling `.start` on it. It needs an instance of a `ContextShift` in order to know which thread pool to run the fiber on. 

Author: ```suggestion
In the previous section, we have seen that fiber runs a set of instructions (the source). Any `IO` can be run in a fiber as long as there is a `ContextShift` type class instance available for it. This is done by calling `.start` on it. It needs an instance of a `ContextShift` in order to know which thread pool to run the fiber on. 
```

Reviewer: 👌

```

Now, let's say we have the following program:
```
val program = for {
  _ <- IO(println(s"What's up?"))
  input <- IO(readLine)
  _ <- IO(println(s"Ah, $input is up!"))
} yield ExitCode.Success
```

Run loop stack for the program above would then look like this:
```
FlatMap(
  FlatMap(
    Delay(() => print("What's up?")),
    (_: Unit) => Delay(() => readLine)
  ),
  input => Delay(() => println(s"Ah, $input is up!"))
)
```
([here](https://github.com/typelevel/cats-effect/blob/master/core/shared/src/main/scala/cats/effect/internals/IORunLoop.scala) is the link to the actual cats-effect run loop)

When the program is run (at the "end of the world", in our case using `unsafeRunSync()`), the stack is submitted to the scheduler. Is it submitted all at once? Or is it submitting one layer of FlatMap at a time, each time yielding back when the task is completed, allowing other fibers to run? 

Actually, this is up to the library / construct being used:

- Scala Future: Yields back on every `FlatMap`
- IO: Yields back when the whole stack is completed
- Monix Task: Yields back every N operations (at the moment of writing this, I believe N = 1024).

> This means that Scala `Future` optimizes for fairness, `IO` for throughput, and `Monix` takes the middle approach. 

Author: Good point!

Note that we could prevent Future from yielding all the time by using a non-shifting instance of `ExecutionContext`, and we could force `IO` to shift manually when we want it to. 
You can think of Scala Future as being "fairness opt-out" and IO "fairness opt-in".

## Cooperative yielding

The relationship between fibers and threads is the following:
- It is possible to have each fiber running on its dedicated thread
- It is possible to have all fibers running on only one thread

> - It is possible to have one fiber switching between mutliple threads

Author: ```suggestion
- It is possible to have one fiber switching between multiple threads
```

Because the number of fibers is usually higher than the number of threads in the given thread pool, fibers yield control to each other in order to make sure that all fibers get their piece of the CPU time. 
For example, in a pool with two threads that's running three fibers, one fiber will be waiting at any given point. 

In cats-effect 2, cooperative yielding is controlled via `ContextShift`. 
Note that `ContextShift` is most likely going to be removed in cats-effect 3; see [ContextShift in cats-effect 3](#contextshift-in-cats-effect-3).

### ContextShift

(Note: `ContextShift` is [getting removed in cats-effect 3](#cats-effect-3), but core principles explained here are still useful to understand, because they will still be relevant in the next version)

In cats-effect, submitting the fiber to a thread pool is done via `ContextShift` type. It has two main abilities: to run the continuation on some `ExecutionContext`, and to shift it to a different `ExecutionContext`.

Here's the trait:

```
trait ContextShift[F[_]] {
  def shift: F[Unit]
  def evalOn[A](ec: ExecutionContext)(f: F[A]): F[A]
}
```

> You can think of it as a type class, even though it is not really a valid type class because it doesn't have the coherence restriction - a type can implement a type class in more than one way. For example, you might want to have a bunch of instances of `ContextShift[IO]` lying around, each constructed using a different `ExecutionContext` and representing a different thread pool (one for blocking I/O, one for CPU-heavy stuff, etc.). As mentioned wearlier, constructing an instance of `ContextShift[IO]` is easy: `val cs = IO.contextShift(executionContext)`.

Author: ```suggestion
You can think of it as a type class, even though it is not really a valid type class because it doesn't have the coherence restriction - a type can implement a type class in more than one way. For example, you might want to have a bunch of instances of `ContextShift[IO]` lying around, each constructed using a different `ExecutionContext` and representing a different thread pool (one for blocking I/O, one for CPU-heavy stuff, etc.). As mentioned earlier, constructing an instance of `ContextShift[IO]` is easy: `val cs = IO.contextShift(executionContext)`.
```

Reviewer: Wait, you don't know the word "wearlier"?

Method `shift` is how we achieve fairness. 
Every fiber will be executed synchronously until shifted, at which point other fibers have the chance to advance their work. 

Don't confuse `shift` from `ContextShift` with `IO.shift`. 
The semantics are the same, but they come in slightly different forms. 
`IO` version has the following two overloads of `shift` method:

```
def shift(implicit cs: ContextShift[IO]): IO[Unit]
def shift(ec: ExecutionContext): IO[Unit]
```

These two methods are similar in nature - they both shift to the desired thread pool, one by providing the Scala's `ExecutionContext`, the other one by providing a `ContextShift`. It is recommended to use `ContextShift` by default, and to provide `ExecutionContext` only when you need fine-grained control over the thread pool in use. Note that you can simply provide the same `ContextShift` / `ExecutionContext` that you're already running on, which will have the effect of cooperatively yielding to other fibers on the same thread pool, same as `shift` from the type class (you can even invoke it simply as `IO.shift`, as long as you have your `ContextShift` available implicitly). 

> ### Examples

Author: Awesome so far!!
Progress checkpoint

Reviewer: Thanks! The whole thing around `ContextShift` is a bit hard to explain, with `shift` doing several things at once, and being present in `ContextShift`, `Async` and `IO`. I think it's a great improvement that we're getting rid of all this in ce3 (also explained in the text).

```
  val program = for {
    _ <- loop("A")(0) // .start(cs)
    _ <- loop("B")(0) // .start(cs)
  } yield ExitCode.Success
```
Homework: Try to run the example without `.start`, that is, without spawning any separate fibers, but keep the `shift` inside the loop. What happens? What gets printed out?

### Summary

Type class `ContextShift` gives us the ability to execute an effect on some desired thread pool via `evalOn` by providing the corresponding `ExecutionContext`, or to `shift`, which re-schedules the fiber in the current thread pool and enables cooperative yielding. 

We get the same two abilities in `IO`, but in that case `shift` takes the thread pool as a parameter, either as `ExecutionContext` or (implicit) `ContextShift`. 
This means that we can cooperatively yield to another fiber within the same thread pool by passing the reference to the current one, and we can also shift to a different one.

In the case of `Future`, there are no fibers. We pass an `ExecutionContext` to each map / flatMap call, which means that every `Future` computation might be executed on a different thread (this is up to the passed `ExecutionContext` and how it decides to schedule the work).

> What we cannot do with `Future`s, however, is define two concurrent computations that will reuse the same thread cooperatively.

Author: 👍 👍 👍

Reviewer: 🕺

- Erlang [processes](https://www.tutorialspoint.com/erlang/erlang_processes.htm) 
- Julia [Tasks](https://docs.julialang.org/en/v1/base/parallel/)
- Common Lisp via [green-threads](https://github.com/thezerobit/green-threads) library
- And many others

## References

- Cats-effect documentation: https://typelevel.org/cats-effect/  
- Cats-effect repo: https://github.com/typelevel/cats-effect
- Cats-effect 3 proposal: https://github.com/typelevel/cats-effect/issues/634
- Monix Task documentation: https://monix.io/docs/2x/eval/task.html  
- Fabio Labella - How do Fibers work: https://www.youtube.com/watch?v=x5_MmZVLiSM  
- Pawel Jurczenko - Modern JVM Multithreading: https://pjurczenko.github.io/modern-jvm-multithreading.html 
- Daniel Spiewak's gist: https://gist.github.com/djspiewak/46b543800958cf61af6efa8e072bfd5c  
- Java Executors: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Executors.html
- Fork Join Pool vs Thread Pool Executor: http://www.h-online.com/developer/features/The-fork-join-framework-in-Java-7-1762357.html
- Adam Warski about Loom: https://blog.softwaremill.com/will-project-loom-obliterate-java-futures-fb1a28508232?gi=c5487dba95ec

> - Loom-fiber repo: https://github.com/forax/loom-fiber

Author: I'm here! 👍
